Introducing CLI tv-casting-app for Linux

sharadb-amazon opened the pull request:
#### Problem

The CHIP SDK did not have an app that implemented the flow where a commissionable node discovers a commissioner and requests it for commissioning. See [InitiatingSetup.adoc#24-commissioner-discovery-from-an-on-network-device](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/rendezvous/InitiatingSetup.adoc#24-commissioner-discovery-from-an-on-network-device)  

What is being fixed?  
Added a CLI based CHIP TV Casting app for Linux with the following steps implemented (from the flow referenced above): 
* discovers TVs on the local network that act as commissioners
* lets the user select one commissioner TV
* Enters commissioning mode and opens commissioning window
* Advertises itself as a commissionable node
* sends the TV a User Directed Commissioning request to the selected commissioner TV
* Displays the onboarding payload to the user

_TBD: Implementation to actually send a casting command to a TV will be done in a separate PR._

#### Change overview
* New tv-casting-app example (which will act as the commissionable node)
* New RequestCommissioning method in the CHIPCommissionableNodeController
* tv-casting-app leverages the CHIPCommissionableNodeController to execute all the steps mentioned in the above section

#### Testing
A. Test cases added for commissioner discovery

B. Tested the tv-casting-app on Linux (VSCode/docker) by using the tv-app, minimal-mdns-client and chipTool:
1. Started running the tv-app and then the tv-casting-app in parallel, which discovers the commissioner TV over DNS-SD as follows.
$ `out/host/chip-tv-casting-app `

> ...
> [1628206267.250223][40533:40533] CHIP:ZCL: Discovered Commisioner #1
> [1628206267.250266][40533:40533] CHIP:DIS: Device Name: Test TV
> [1628206267.250282][40533:40533] CHIP:DIS: Vendor ID: 9050
> [1628206267.250324][40533:40533] CHIP:DIS: Product ID: 65279
> [1628206267.250340][40533:40533] CHIP:DIS: Device Type: 35
> [1628206267.250381][40533:40533] CHIP:DIS: Hostname: 025000000001
> [1628206267.250397][40533:40533] CHIP:DIS: Port: 5540
> [1628206267.250439][40533:40533] CHIP:DIS: Number of IP addresses: 2. IP Adddress(es): 
> [1628206267.250458][40533:40533] CHIP:DIS: fe80::50:ff:fe00:1, 
> [1628206267.250502][40533:40533] CHIP:DIS: 192.168.65.3, 
> [1628206267.250517][40533:40533] CHIP:DIS: Commissioning Mode: 0
> [1628206267.250561][40533:40533] CHIP:ZCL: 1 commissioner(s) discovered

2. User selects a commissioner TV over the CLI.
> [1628206267.250577][40533:40533] CHIP:ZCL: Choose a commissioner TV (by number# above) to request commissioning from: 
> 1

3. Opens commissioning window, enters commissioning mode and advertises itself as a commissionable node over DNS-SD (verified advertisement with `out/host/minimal-mdns-client -q _matterc._udp.local`)

> [1628194259.587022][11128:11128] CHIP:ZCL: Using ZAP configuration...
> _...snip..._
> [1628194259.595609][11128:11128] CHIP:IN: UDP::Init bind&listen port=5540
> [1628194259.595678][11128:11128] CHIP:IN: UDP::Init bind&listen port=5540
> _...snip..._
> [1628194259.605502][11128:11128] CHIP:SC: Waiting for PBKDF param request
> [1628194259.605533][11128:11128] CHIP:DIS: Start dns-sd server
> [1628194259.606382][11128:11128] CHIP:DIS: CHIP minimal mDNS started advertising.
> [1628194259.606514][11128:11128] CHIP:DIS: Replying to DNS-SD service listing request
> _...snip..._
> [1628194259.609278][11128:11128] CHIP:SVR: Server Listening...

4. Sends User directed commissioning request to the selected commissioner TV (verified UDC commands were received on the tv-app console)

> [1628194259.609290][11128:11128] CHIP:SVR: SendUserDirectedCommissioningRequest2
> [1628194259.609301][11128:11128] CHIP:SVR: instanceName=DC514873944A5CFF
> [1628194259.609317][11128:11128] CHIP:IN: Sending UDC msg
> [1628194259.610781][11128:11128] CHIP:IN: UDC msg send status ../../third_party/connectedhomeip/src/inet/IPEndPointBasis.cpp:915: No Error
> [1628194259.610850][11128:11128] CHIP:SVR: Send UDC request success
> _...snip..._

5. Displays the onboarding payload
> [1628206286.424116][40533:40533] CHIP:ZCL: Onboarding Payload::
> [1628206286.424207][40533:40533] CHIP:ZCL: PIN: 34567890
> [1628206286.424225][40533:40533] CHIP:ZCL: Discriminator: 2976
> [1628206286.424253][40533:40533] CHIP:ZCL: Vendor ID: 9050
> [1628206286.424306][40533:40533] CHIP:ZCL: Product ID: 65279

6. Receives onnetwork commissioning command from chipTool (`out/host/chip-tool pairing onnetwork 0 34567890 2976 192.168.65.4 5540`) and gets commissioned successfully.

> [1628194320.944094][11128:11128] CHIP:EM: Received message of type 0x20 with vendorId 0x0000 and protocolId 0x0000 on exchange 39017
> _...snip..._
> [1628194320.956272][11128:11128] CHIP:SC: Sent PBKDF param response
> [1628194320.967469][11128:11128] CHIP:EM: Received message of type 0x22 with vendorId 0x0000 and protocolId 0x0000 on exchange 39017
> _...snip..._
> [1628194320.971048][11128:11128] CHIP:SC: Sent spake2p msg2
> [1628194320.979742][11128:11128] CHIP:EM: Received message of type 0x24 with vendorId 0x0000 and protocolId 0x0000 on exchange 39017
> _...snip..._
> [1628194321.277809][11128:11128] CHIP:IN: CASE Session established. Setting up the secure channel.
> [1628194321.280133][11128:11128] CHIP:IN: Marking old secure session for device 0x000000000001B669 as expired
> [1628194321.280173][11128:11128] CHIP:IN: New secure session created for device 0x000000000001B669, key 254!!
> [1628194321.280251][11128:11128] CHIP:IN: CASE secure channel is available now.
> [1628194321.280269][11128:11128] CHIP:IN: CASE Server enabling CASE session setups
> [1628194321.280298][11128:11128] CHIP:EM: Sending Standalone Ack for MsgId:00000000
> [1628194321.280446][11128:11128] CHIP:EM: Flushed pending ack for MsgId:00000000
> [1628194444.416157][11128:11128] CHIP:DL: Inet Layer shutdown
> [1628194444.416263][11128:11128] CHIP:DL: BLE layer shutdown
> [1628194444.416311][11128:11128] CHIP:DL: System Layer shutdown

andy31415 commented:
Could you separate out commits between generated files and hand-edited code?

This is a 12K line added PR and it would help a lot to have the 'generated files' separated out.

sharadb-amazon commented:
> Could you separate out commits between generated files and hand-edited code?
> 
> This is a 12K line added PR and it would help a lot to have the 'generated files' separated out.

Sure, I just separated the changes into different commits (zap/gen code and handwritten code): https://github.com/project-chip/connectedhomeip/pull/8824/commits - is that better?

chrisdecenzo commented on BUILD.gn:
```diff
@@ -435,6 +447,9 @@ if (current_toolchain != "${dir_pw_toolchain}/default:default") {
       deps += [ ":linux_tv_app" ]
     }
+    if (enable_linux_tv_casting_app_build) {
+      deps += [ ":linux_tv_app" ]
```
Not sure what this line does but should this be the following:

```suggestion
      deps += [ ":linux_tv_casting_app" ]
```

Wondering if this was a copy/paste error...

sharadb-amazon replied:
Ah.. yes. Thanks for catching that.

chrisdecenzo commented on src/lib/mdns/Resolver.h:
```diff
@@ -100,6 +102,66 @@ struct DiscoveredNodeData
+        {
+            char rotatingIdString[chip::Mdns::kMaxRotatingIdLen * 2 + 1] = "";
+            Encoding::BytesToUppercaseHexString(rotatingId, rotatingIdLen, rotatingIdString, sizeof(rotatingIdString));
+            printf("Rotating ID: %s\n", rotatingIdString);
```
should these be ChipLogInfo calls rather than printf?

sharadb-amazon replied:
Switched to ChipLogProgress.

chrisdecenzo reviewed (approved):
please see comments

bzbarsky-apple commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,106 @@
+    // Display discovered commissioner TVs
+    for (int i = 0; i < CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES; i++)
+    {
+        const Mdns::DiscoveredNodeData * commissioner = mCommissionableNodeController.GetDiscoveredCommissioner(i);
```
Sure seems like we should just have a way to enumerate discovered commissioners on the controller, instead of looping over indices in a way that is totally undocumented (there is zero documentation for `CommissionableNodeController::GetDiscoveredCommissioner`) and so on.  Please at the very least add clear documentation about proper usage of `GetDiscoveredCommissioner` and what a null return from it does and does not mean and so on.

sharadb-amazon replied:
Added documentation to GetDiscoveredCommissioner

bzbarsky-apple commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
+            break;
+        }
+    }
+    if (commissionerNumber != 0)
```
This does not seem correct to me.

Let's say we have discovered absolutely nothing and the caller passes in `0` as `commissionerNumber`.  We'll happily proceed past here and proceed to crash on a null-deref.

sharadb-amazon replied:
If we discover absolutely nothing, we will only display that 0 commissioners were found and exit (see line #80 VerifyOrExit)

bzbarsky-apple replied:
My point is that this function's safety depends on something non-obvious being done by its callers.  Nothing documents that you MUST NOT call this function if nothing was discovered and that it will crash if you do.  The use right now happens to be safe, but it's way too easy for the caller to change and become unsafe.

It would be better if this function simply didn't do unsafe things no matter what the caller passed in and under what conditions.

bzbarsky-apple commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
+    }
+
+    // Enter commissioning mode, open commissioning window
+    InitServer();
```
This also does not look right to me.  The "server" should not be controlled by this random component; it's a shared service.  The app should be doing the `InitServer` bits.

sharadb-amazon replied:
Moved calls to InitServer, and others to app/server class methods to example app main.cpp

bzbarsky-apple commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
+
+    // Send User Directed commissioning request
+    chip::Transport::PeerAddress commissionerAddress =
+        chip::Transport::PeerAddress::UDP(commissioner->ipAddress[0], commissioner->port);
```
This needs an interface id too, if we discovered a link-local ipv6 address.  `PacketDataReporter` should be storing that in the `DiscoveredNodeData` like it does in the `ResolvedNodeData`, no?

Please make sure to test this scenario.

sharadb-amazon replied:
Stored the interfaceIds

bzbarsky-apple commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
+    InitServer();
+
+    // Send User Directed commissioning request
+    chip::Transport::PeerAddress commissionerAddress =
```
```suggestion
    auto commissionerAddress =
```

bzbarsky-apple commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
+        chip::Transport::PeerAddress::UDP(commissioner->ipAddress[0], commissioner->port);
+    SendUserDirectedCommissioningRequest(commissionerAddress);
+#else
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+#endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+    return CHIP_NO_ERROR;
```
The `#else` case here is liable to complain about unreached code on some compilers.  Shouldn't the `return CHIP_NO_ERROR` move into the `#if` body?

bzbarsky-apple commented on src/controller/CHIPCommissionableNodeController.h:
```diff
@@ -54,6 +55,10 @@ class DLL_EXPORT CommissionableNodeController : public AbstractMdnsDiscoveryCont
         ChipLogError(Controller, "Unsupported operation CommissionableNodeController::OnNodeIdResolutionFailed");
     }
 
+    CHIP_ERROR RequestCommissioning(int commissionerIdx);
```
Document, please.  This does not have trivial semantics in terms of what the argument means.

bzbarsky-apple commented on src/controller/CHIPCommissionableNodeController.h:
```diff
@@ -54,6 +55,10 @@ class DLL_EXPORT CommissionableNodeController : public AbstractMdnsDiscoveryCont
 
+    CHIP_ERROR RequestCommissioning(int commissionerIdx);
+
+    CHIP_ERROR GetOnboardingPayload(chip::SetupPayload & aOnboardingPayload);
```
```suggestion
    CHIP_ERROR GetOnboardingPayload(SetupPayload & aOnboardingPayload);
```

bzbarsky-apple commented on src/lib/mdns/Discovery_ImplPlatform.cpp:
```diff
@@ -477,6 +477,8 @@ void DiscoveryImplPlatform::HandleNodeResolve(void * context, MdnsService * resu
         data.ipAddress[data.numIPs++] = result->mAddress.Value();
     }
 
+    data.port = result->mPort;
```
Need to store the interface id somewhere here too.

sharadb-amazon replied:
Done.

bzbarsky-apple commented on src/lib/mdns/Resolver_ImplMinimalMdns.cpp:
```diff
@@ -161,6 +161,7 @@ void PacketDataReporter::OnCommissionableNodeSrvRecord(SerializedQNameIterator n
     {
         strncpy(mDiscoveredNodeData.instanceName, name.Value(), sizeof(DiscoveredNodeData::instanceName));
     }
+    mDiscoveredNodeData.port = srv.GetPort();
```
Need to store the interface id somewhere in this general code too.

sharadb-amazon replied:
Done.

tcarmelveilleux commented on src/app/server/OnboardingCodesUtil.cpp:
```diff
@@ -113,7 +115,8 @@ void ShareQRCodeOverNFC(chip::RendezvousInformationFlags aRendezvousFlags)
 {
-    CHIP_ERROR err                      = CHIP_NO_ERROR;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+#if CONFIG_DEVICE_LAYER
```
Why not refactor and raise issues to avoid a dependency that seems wrong, and breaks the example you are working on, rather than adding more #ifdefs that just push the problem further down?

sharadb-amazon replied:
No changes to OnboardingCodesUtil or other classes under app/server in the updated diff.

tcarmelveilleux commented on src/app/server/RendezvousServer.cpp:
```diff
@@ -36,15 +39,18 @@ using namespace ::chip::DeviceLayer;
 namespace chip {
 
 namespace {
+#if CONFIG_DEVICE_LAYER
```
Please express the root cause of needing these new `#ifdef`. If it is that we have incorrect APIs for your usecases, then others will suffer too, and we need to converge/fix, rather than hard-to-maintain and not trivial ifdefs.

sharadb-amazon replied:
I had to put them in because I was creating a build dependency from the controller to app/server. When I moved calls to InitServer (and then to other app/server functions to the example app main.cpp function), I did not need those #ifdefs. The latest diff does not touch any of the classes in app/server.

tcarmelveilleux commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
+    for (int i = 0; i < CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES; i++)
+    {
+        commissioner = GetDiscoveredCommissioner(i);
+        if (commissioner != nullptr && --commissionerNumber == 0)
```
`--commissionerNumber == 0` is not an obvious condition. What is being done here?

sharadb-amazon replied:
Added documentation inline in the code.

tcarmelveilleux commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
 
+CHIP_ERROR CommissionableNodeController::RequestCommissioning(int commissionerNumber)
+{
+#if CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
```
Why this ifdef if the command only applyes to Commissioner discovery? It seems like the API should just not exist in the parent class in some examples. Appears to me to be a good example of needing API design and possibly introducing classes for your usecase

tcarmelveilleux commented on src/controller/CHIPCommissionableNodeController.cpp:
```diff
@@ -35,5 +37,39 @@ const Mdns::DiscoveredNodeData * CommissionableNodeController::GetDiscoveredComm
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    // Enter commissioning mode, open commissioning window
```
The fact that `InitServer()` opens a commissioning window is temporary behavior, and when spec compliance is improved with regards to commissioning window, it's possible no commissioning window will be open at all from this call. If you desire to forcible open a commissioning window, there are other means.

sharadb-amazon replied:
Called OpenBasicCommissioningWindow() to ensure a commissioning window is opened

tcarmelveilleux commented on src/lib/mdns/Resolver.h:
```diff
@@ -100,6 +102,65 @@ struct DiscoveredNodeData
+        {
+            ChipLogProgress(Discovery, "Additional Pairing: %u", additionalPairing);
+        }
+        if (strcmp(pairingInstruction, "") != 0)
```
Suggest if (strlen(pairingInstruction) == 0)

sharadb-amazon replied:
Done.

sharadb-amazon replied:
Done.

tcarmelveilleux commented on src/lib/mdns/Resolver.h:
```diff
@@ -100,6 +102,65 @@ struct DiscoveredNodeData
     DiscoveredNodeData() { Reset(); }
     bool IsHost(const char * host) const { return strcmp(host, hostName) == 0; }
     bool IsValid() const { return !IsHost("") && ipAddress[0] != chip::Inet::IPAddress::Any; }
+    void Log() const
```
This may cause sufficient logging to overflow some logs in case of large number of nodes present (e.g. 100+). The method should not use `ChipLogProgress`, but rather `ChipLogDetail` (which is the verbose logging level)

Furthermore, a new logging module like `DiscoveryVerbose` could be added, so that it can be disabled if needed. This will avoid embedded targets from all using this code by default, and building it by default.

This is used only by your new code, but added in a way that makes it enticing to use during every discovery, which may/will impact performance on some targets if used more widely, especially in presence of large number of nodes.

Suggest overall to replace the Log() method here with a `Log(const DiscoveredNodeData &)` in your own example/application, since the format seems specific to the usecase here.

sharadb-amazon replied:
Note that there are other places in the code that can reduce repetition by just calling this function. Here is one: https://github.com/project-chip/connectedhomeip/blob/master/src/controller/python/ChipDeviceController-ScriptBinding.cpp#L336-L367

Agreed with the first couple of suggestions here: Switched to ChipLogDetail and the new DiscoveryVerbose module

tcarmelveilleux commented on src/app/server/Mdns.cpp:
```diff
@@ -93,13 +98,15 @@ CHIP_ERROR GetCommissionableInstanceName(char * buffer, size_t bufferLen)
 /// Set MDNS operational advertisement
 CHIP_ERROR AdvertiseOperational()
 {
-    uint64_t fabricId;
+    uint64_t fabricId = 5544332211; // default value
 
+#if CONFIG_DEVICE_LAYER
```
The AdvertiseOperational may be too "cast" to devices (and require direct access to underlying storage APIs). It seems like you have found another spot where your usecase is impacted by the current API assumptions. However, rather than raise and discuss how to solve for all cases, there is yet more `#ifdef` here and more assumptions being added.

sharadb-amazon replied:
No changes to Mdns.cpp anymore in the latest diff.

tcarmelveilleux reviewed (changes requested):
While I can understand that the current state of some of the API surfaces make the desired change harder to do, the solution used here to add a lot of application-dependant `#ifdef` will *not* scale. This is just one more application desiring to implement a given feature, and I can imagine N other such applications with slightly different requirements. Instead of adding more `#ifdef` that enshrine more assumptions (some of which are temporary), there should be some effort to refactor or at least discuss the obstacles found in API usage that cause the need to include all this specific steering of code. 

The pervasive new usage of `#if CONFIG_DEVICE_LAYER` ifdefs really indicate to me that some APIs are contradict some the assumptions of the author in terms of what their application ought to have to provide or implement. Rather than `#ifdef`, let's discuss how to improve the API for everyone.

Finally, this PR touches much code, including code not related to its own requirements, and does so *without adding or modifying a single unit test or integration test*. Manual testing of a feature of this size is insufficient to maintain it in the tree.

andy31415 commented:
Flagging as SDK Discussion required concering the approach: sudo usage (is it BT, something else)?, usage of ifdefs, split between functionality and example app  development. Generally looking for a path forward to have incremental development instead of large patches mixing examples and functionalityt code.

sharadb-amazon commented on examples/chip-tool/commands/discover/DiscoverCommissionersCommand.cpp:
```diff
@@ -34,45 +34,11 @@ void DiscoverCommissionersCommand::Shutdown()
-            }
-
-            printf("\n");
+            ChipLogProgress(chipTool, "Discovered Commisioner #%d", ++commissionerCount);
```
Done

sharadb-amazon commented on examples/tv-casting-app/linux/README.md:
```diff
@@ -0,0 +1,53 @@
+-   Pre-requisite: Build and run the tv-app
+
+        $ cd ~/connectedhomeip/examples/tv-app/linux
+        $ sudo out/debug/chip-tv-app
```
sudo is not needed (it was a copy-paste error, copied from tv-app README). Updated the README

sharadb-amazon commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,106 @@
+{
+    CHIP_ERROR err        = CHIP_NO_ERROR;
+    int commissionerCount = 0, selectedCommissioner = CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES;
+    chip::Controller::CommissionableNodeController mCommissionableNodeController;
```
Removed the m initial.

tcarmelveilleux commented on examples/tv-casting-app/linux/third_party/connectedhomeip:
```diff
@@ -0,0 +1 @@
+../../../../
```
This appears to be a symlink. Why?

sharadb-amazon replied:
Same as most other examples (see https://github.com/project-chip/connectedhomeip/blob/master/examples/lighting-app/linux/third_party/connectedhomeip or https://github.com/project-chip/connectedhomeip/blob/master/examples/all-clusters-app/linux/third_party/connectedhomeip). I think it is for convenience? READMEs carry the following activation instruction: "source third_party/connectedhomeip/scripts/activate.sh"

tcarmelveilleux commented on src/lib/mdns/Discovery_ImplPlatform.cpp:
```diff
@@ -474,9 +474,13 @@ void DiscoveryImplPlatform::HandleNodeResolve(void * context, MdnsService * resu
     if (result->mAddress.HasValue())
     {
-        data.ipAddress[data.numIPs++] = result->mAddress.Value();
+        data.ipAddress[data.numIPs]   = result->mAddress.Value();
+        data.interfaceId[data.numIPs] = result->mInterface;
+        data.numIPs++;
```
Could this overrun the array? There are no bound checks on `data.numIPs` versus `ArraySize(data.ipAddress)`, etc

tcarmelveilleux commented on src/lib/mdns/Resolver.h:
```diff
@@ -100,6 +103,66 @@ struct DiscoveredNodeData
     bool IsValid() const { return !IsHost("") && ipAddress[0] != chip::Inet::IPAddress::Any; }
+    void Log() const
+    {
+        if (rotatingIdLen > 0)
+        {
+            char rotatingIdString[chip::Mdns::kMaxRotatingIdLen * 2 + 1] = "";
+            Encoding::BytesToUppercaseHexString(rotatingId, rotatingIdLen, rotatingIdString, sizeof(rotatingIdString));
+            ChipLogDetail(DiscoveryVerbose, "Rotating ID: %s", rotatingIdString);
+        }
```
This must be conditional on the rotating ID feature (which is not always implemented)

sharadb-amazon replied:
The generation and advertisement of a rotating ID is already conditional on the server side on #if CHIP_ENABLE_ROTATING_DEVICE_ID. The client side code here is already checking if no rotating ID was received by checking for rotatingIdLen > 0. Then, why must we put a #if CHIP_ENABLE_ROTATING_DEVICE_ID check here in the client?

tcarmelveilleux replied:
Because rotating ID is a fully optional feature. Try to turn off CHIP_ENABLE_ROTATING_DEVICE_ID and it should build. If it builds cleanly without additional `#ifdef` after `CHIP_ENABLE_ROTATING_DEVICE_ID` is disabled, then you can resolve this comment.

sharadb-amazon replied:
Sure. I defined CHIP_ENABLE_ROTATING_DEVICE_ID and set it to 0 in the tv-casting-app's CHIPProjectAppConfig.h file. The build still succeeded.

tcarmelveilleux commented on src/lib/mdns/Resolver.h:
```diff
@@ -100,6 +103,66 @@ struct DiscoveredNodeData
+            {
+                char ipAddressString[Inet::kMaxIPAddressStringLength];
+                ChipLogDetail(DiscoveryVerbose, "IP Address #%d: %s", j + 1,
+                              ipAddress[j].ToString(ipAddressString, sizeof(ipAddressString)));
+                (void) ipAddressString;
```
Please separate the conversion from the logging, and remove `(void)ipAddressString`

sharadb-amazon replied:
Separated

tcarmelveilleux commented on src/lib/mdns/Resolver_ImplMinimalMdns.cpp:
```diff
@@ -181,7 +184,9 @@ void PacketDataReporter::OnDiscoveredNodeIPAddress(const chip::Inet::IPAddress &
         return;
     }
-    mDiscoveredNodeData.ipAddress[mDiscoveredNodeData.numIPs++] = addr;
+    mDiscoveredNodeData.ipAddress[mDiscoveredNodeData.numIPs]   = addr;
+    mDiscoveredNodeData.interfaceId[mDiscoveredNodeData.numIPs] = mInterfaceId;
+    mDiscoveredNodeData.numIPs++;
```
Same here: how do we guarantee bound checks on the numIPs argument

sharadb-amazon replied:
There is a check before these assignments at line 182.

sharadb-amazon commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,106 @@
+            aDelayMilliseconds,
+            [](System::Layer *, void *, CHIP_ERROR err) {
+                DeviceLayer::PlatformMgr().StopEventLoopTask();
+                DeviceLayer::PlatformMgr().Shutdown();
```
As discussed yesterday, we found a bug in src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp. The documentation is fine I think, I just got led astray by that bug and ended up using these Shutdowns because I could not get it to work any other way (I started thinking that I was misinterpreting the documentation).

I have updated the code now to not do these Shutdowns and will post a separate PR for the fix in the GenericPlatformManagerImpl_POSIX.cpp

sharadb-amazon replied:
Just following up here. Here is the PR for the fix in the GenericPlatformManagerImpl_POSIX : https://github.com/project-chip/connectedhomeip/pull/9094

sharadb-amazon commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,106 @@
+        if (commissioner != nullptr)
+        {
+            ChipLogProgress(Zcl, "Discovered Commisioner #%d", ++commissionerCount);
+            commissioner->Log();
```
Done.

sharadb-amazon commented:
> While I can understand that the current state of some of the API surfaces make the desired change harder to do, the solution used here to add a lot of application-dependant `#ifdef` will _not_ scale. This is just one more application desiring to implement a given feature, and I can imagine N other such applications with slightly different requirements. Instead of adding more `#ifdef` that enshrine more assumptions (some of which are temporary), there should be some effort to refactor or at least discuss the obstacles found in API usage that cause the need to include all this specific steering of code.
> 
> The pervasive new usage of `#if CONFIG_DEVICE_LAYER` ifdefs really indicate to me that some APIs are contradict some the assumptions of the author in terms of what their application ought to have to provide or implement. Rather than `#ifdef`, let's discuss how to improve the API for everyone.
> 
> Finally, this PR touches much code, including code not related to its own requirements, and does so **without adding or modifying a single unit test or integration test**. Manual testing of a feature of this size is insufficient to maintain it in the tree.

As mentioned in reply to one of the inline comments, the ifdefs were required when functions in app/server were being called from the controller. I refactored the code and these ifdefs are not needed anymore. 

Also, test cases have been added for commissioner discovery.

sharadb-amazon commented:
> Flagging as SDK Discussion required concering the approach: sudo usage (is it BT, something else)?, usage of ifdefs, split between functionality and example app development. Generally looking for a path forward to have incremental development instead of large patches mixing examples and functionality code.

Discussed with tcarmelveilleux@ and bzbarsky-apple@ on Slack since we did not have the SDK discussion last week. Addressed comments - sudo usage, ifdefs not needed anymore - diff has been updated.

andy31415 commented on src/lib/support/logging/CHIPLogging.cpp:
```diff
@@ -91,6 +91,7 @@ static const char ModuleNames[] = "-\0\0" // None
                                   "SPL"   // SetupPayload
                                   "SVR"   // AppServer
                                   "DIS"   // Discovery
+                                  "DSV"   // DiscoveryVerbose
```
Why not ChipLogDetail ?

This seems to set an unusual precedent: having some verbosity flags as a separate module.

sharadb-amazon replied:
The rationale for this comes from one of Tennessee's previous comments. Copying here: "_Furthermore, a new logging module like DiscoveryVerbose could be added, so that it can be disabled if needed. This will avoid embedded targets from all using this code by default, and building it by default._"

sharadb-amazon replied:
After discussing with andy31415 on Slack, removed the DiscoveryVerbose logging module.

andy31415 commented:
Removed discussion required flag based on review from tennessee and that we do not use ifdef anymore.

Leaving google-side checkmark review up to @mspang as this contains build system changes

xylophone21 commented:
Does that have something different with chip-device-ctrl > discover -all ?

chip-device-ctrl > discover -all
Waiting for device responses...
[1629097627.796048][16287:16295] CHIP:DL: Avahi browse: cache exhausted
[1629097628.797996][16287:16295] CHIP:DL: Avahi browse: all for now
Waiting for device responses...
[1629097629.801754][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.802156][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.802348][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.802756][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.803040][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.803203][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.803516][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.803656][16287:16295] CHIP:DL: Avahi browse: cache new
[1629097629.803773][16287:16295] CHIP:DL: Avahi browse: cache exhausted
[1629097629.803885][16287:16295] CHIP:DL: Avahi browse: all for now
[1629097629.814367][16287:16295] CHIP:DL: Avahi resolve found
[1629097629.815170][16287:16295] CHIP:DL: Avahi resolve found
[1629097629.815937][16287:16295] CHIP:DL: Avahi resolve found
[1629097629.816787][16287:16295] CHIP:DL: Avahi resolve found
[1629097631.793379][16287:16287] CHIP:DIS: Commissionable Node 0
[1629097631.793469][16287:16287] CHIP:DIS: 	Host name:		Android
[1629097631.793542][16287:16287] CHIP:DIS: 	Long discriminator:	0
[1629097631.793613][16287:16287] CHIP:DIS: 	Vendor ID:		0
[1629097631.793691][16287:16287] CHIP:DIS: 	Product ID:		0
[1629097631.793761][16287:16287] CHIP:DIS: 	Additional Pairing	0
[1629097631.793837][16287:16287] CHIP:DIS: 	Commissioning Mode	0
[1629097631.793904][16287:16287] CHIP:DIS: 	Device Type		0
[1629097631.793973][16287:16287] CHIP:DIS: 	Device Name		
[1629097631.794039][16287:16287] CHIP:DIS: 	Rotating Id		
[1629097631.794097][16287:16287] CHIP:DIS: 	Pairing Instruction	
[1629097631.794156][16287:16287] CHIP:DIS: 	Pairing Hint		0x0
[1629097631.794241][16287:16287] CHIP:DIS: 	Address 0:		192.168.3.198
[1629097631.794325][16287:16287] CHIP:DIS: Commissionable Node 1
[1629097631.794372][16287:16287] CHIP:DIS: 	Host name:		DCA63279BEFB
[1629097631.794441][16287:16287] CHIP:DIS: 	Long discriminator:	3840
[1629097631.794500][16287:16287] CHIP:DIS: 	Vendor ID:		9050
[1629097631.794574][16287:16287] CHIP:DIS: 	Product ID:		65279
[1629097631.794637][16287:16287] CHIP:DIS: 	Additional Pairing	0
[1629097631.794710][16287:16287] CHIP:DIS: 	Commissioning Mode	1
[1629097631.794769][16287:16287] CHIP:DIS: 	Device Type		0
[1629097631.794839][16287:16287] CHIP:DIS: 	Device Name		
[1629097631.794897][16287:16287] CHIP:DIS: 	Rotating Id		00007C5F6E176CD40F68685D100A1CF8A98B
[1629097631.794963][16287:16287] CHIP:DIS: 	Pairing Instruction	
[1629097631.795023][16287:16287] CHIP:DIS: 	Pairing Hint		0x21
[1629097631.795098][16287:16287] CHIP:DIS: 	Address 0:		192.168.3.176

xylophone21 commented:
> Does that have something different with chip-device-ctrl > discover
> 
> ## In TE4_TestPlanVerificationSteps_V5, G34
> Put the device in commisioning mode, Verify device advertising its ability for commissioning by running 'discover' command in the python controller along with the QR code or '-v vendorID' .. etc look for discover command usage and pass appropriate param to have commissioner discover the commissionable node . Look for the device information in the below command.
> 
> chip-device-ctrl > discover -qr MT:YNJV7VSC00KA0648G00
> Waiting for device responses...
> [1626844356.667520][1256856:1256864] CHIP:DL: Avahi browse: cache exhausted
> [1626844356.770602][1256856:1256864] CHIP:DL: Avahi browse: cache new
> [1626844356.771830][1256856:1256864] CHIP:DL: Avahi browse: cache new
> [1626844357.669397][1256856:1256864] CHIP:DL: Avahi browse: all for now
> [1626844357.879577][1256856:1256864] CHIP:DL: Avahi resolve found
> [1626844357.881913][1256856:1256864] CHIP:DL: Avahi resolve found
> [1626844358.060376][1256856:1256856] CHIP:DIS: Commissionable Node 0
> [1626844358.060452][1256856:1256856] CHIP:DIS: Host name: DCA6328D7072
> [1626844358.060526][1256856:1256856] CHIP:DIS: Long discriminator: 3840
> [1626844358.060591][1256856:1256856] CHIP:DIS: Vendor ID: 9050
> [1626844358.060656][1256856:1256856] CHIP:DIS: Product ID: 65279
> [1626844358.060748][1256856:1256856] CHIP:DIS: Additional Pairing 0
> [1626844358.060808][1256856:1256856] CHIP:DIS: Commissioning Mode 1
> [1626844358.060884][1256856:1256856] CHIP:DIS: Device Type 257
> [1626844358.060950][1256856:1256856] CHIP:DIS: Device Name Test Bulb
> [1626844358.061028][1256856:1256856] CHIP:DIS: Rotating Id 00007C5F6E176CD40F68685D100A1CF8A98B
> [1626844358.061089][1256856:1256856] CHIP:DIS: Pairing Instruction
> [1626844358.061160][1256856:1256856] CHIP:DIS: Pairing Hint 0x21
> [1626844358.061219][1256856:1256856] CHIP:DIS: Address 0: 192.168.4.223
> chip-device-ctrl >

I see, this feature is not supported by chip-tool, but do we  really need a standalone tv-casting-app? Or cloud added a command in chip tool?

mspang commented on src/BUILD.gn:
```diff
@@ -92,6 +93,11 @@ if (chip_build_tests) {
       deps += [ "${chip_root}/src/ble/tests" ]
     }
 
+    if (!is_clang && chip_device_platform != "darwin" &&
```
Why doesn't it work with gcc?

sharadb-amazon replied:
Opened an issue related to this: https://github.com/project-chip/connectedhomeip/issues/9173

msandstedt commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,108 @@
+    if (commissionerCount > 0)
+    {
+        ChipLogProgress(Zcl, "Choose a commissioner (by number# above) to request commissioning from: ");
+        scanf("%d", &selectedCommissionerNumber);
```
Capturing some discussion that occurred over slack.

The app is now mostly restructured to operate in a single event loop.  However, this scanf capture of user input is still blocking.

One way to solve this is to have the system layer asynchronously watch stdin on behalf of the app, with the result that this method, `RequestCommissioning`, would be split in two: the first half would execute as it does here by callback from ScheduleWork.

The second half would be triggered as one or more callbacks when read of stdin unblocks, until enough user input is provided for the second half of this method to execute.  With these changes, I believe the app would be fully async.

@kpschoedel provided these pre and post #9086 recipes to do this:

```
// without #9086
System::WatchableSocket t;
t.Init(sl.WatchableEventsManager());
t.Attach(fd);
t.SetCallback(callback, data);
t.RequestCallbackOnPendingRead();
// and when done
t.ReleaseFD()
```

```
// with #9086
System::Layer as sl:
SocketWatchToken t;
sl.StartWatchingSocket(fd, &t);
slSetCallback(t, callback, data);
sl.RequestCallbackOnPendingRead(t);
sl.StopWatchingSocket(&t);
```

I am not personally insisting that this has to happen before merge.  Maybe it's arguable it could be a TODO.

msandstedt commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,140 @@
+                       ChipLogError(Zcl, "Could not set non-blocking mode for user input!"));
+        ReturnOnFailure(chip::DeviceLayer::SystemLayer.StartWatchingSocket(0, &token));
+        ReturnOnFailure(chip::DeviceLayer::SystemLayer.SetCallback(token, RequestUserDirectedCommisisioning,
+                                                                   reinterpret_cast<intptr_t>(&commissionerCount)));
```
The passed data happens not to be used in `RequestUserDirectedCommisisioning`.  Still, it is incorrect to pass a pointer to commissionerCount, which is on the stack.

```suggestion
                                                                   nullptr));
```

sharadb-amazon replied:
Thanks for catching this. Actually we don't need any data passed to the callback in this case. So I have updated the diff to just pass in a null.

msandstedt reviewed (approved):
Minor nit inline, but this looks good to me.

tcarmelveilleux commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,139 @@
+ * Accepts user input of selected commissioner and calls PrepareForCommissioning with
+ * the selected commissioner
+ */
+void RequestUserDirectedCommisisioning(System::SocketEvents events, intptr_t data)
```
nit: Typo in Commissioning

tcarmelveilleux commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,139 @@
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Failed to run TV Casting App:: %s", ErrorStr(err));
```
nit: double comma

tcarmelveilleux commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,139 @@
+            Zcl, "%d commissioner(s) discovered. Select one (by number# above) to request commissioning from: ", commissionerCount);
+
+        // Setup for async/non-blocking user input from stdin
+        int flags = fcntl(0, F_GETFL, 0);
```
Suggest using `STDIN_FILENO` instead of `0` as fd fpr fcntl

tcarmelveilleux commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,139 @@
+{
+    // Accept user selection for commissioner to request commissioning from
+    int selectedCommissionerNumber = CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES;
+    scanf("%d", &selectedCommissionerNumber);
```
What happens here if data not available due to non-blocking? Does non-blocking stdin actually work against stdin? This will unblock on socket watching, but scanf() may block if there had been insufficient input. Is it `\n` that unblocks stdin in this case?

msandstedt replied:
I had the same question.  Yes, we could use raw mode and poll each character.  But I'm pretty sure that the kernel will have line buffering in most environments by default, so in a reasonable majority of cases stdin will only unblock when the user actually presses enter.

tcarmelveilleux replied:
OK, then I suggest making that assumption (obviously made here) clear with a comment.

mspang replied:
Unresolving this.. I really don't think we should land examples that block the main thread with interactive input.

mspang replied:
I see that has been addressed although it it not obvious and I think there are still issues with it

- direct scanf won't well with more than one command
- end of file results in an infinite loop except on a tty (e.g. suppose this example is run as a subprocess with input redirected. It will not function)

Overall I wouldn't recommend this approach. If we want line editing in examples we should do it via shell_common.

andy31415 commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,141 @@
+using namespace chip;
+using namespace chip::Controller;
+
+#define TV_DEVICE_TYPE 35
```
please use C++ constants: `constexpr int kTvDeviceType = 35`. This does type checks and compiler will also warn of unused constants if this becomes obsolete.

Should this constant not come from some other include?

andy31415 commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,141 @@
+{
+    // Enter commissioning mode, open commissioning window
+    InitServer();
+    ReturnOnFailure(OpenBasicCommissioningWindow(ResetFabrics::kYes, 3 * 60));
```
use  a constant for the 3*60 value, to have a name for it.

andy31415 commented on examples/tv-casting-app/linux/main.cpp:
```diff
@@ -0,0 +1,141 @@
+
+    // Give commissioners some time to respond and then ScheduleWork to initiate commissioning
+    DeviceLayer::SystemLayer.StartTimer(
+        5 * 1000, [](System::Layer *, void *) { chip::DeviceLayer::PlatformMgr().ScheduleWork(InitCommissioningFlow); }, nullptr);
```
add constant for 5*1000

bzbarsky-apple commented on src/lib/mdns/Resolver.h:
```diff
@@ -138,6 +141,66 @@ struct DiscoveredNodeData
+        for (int j = 0; j < numIPs; j++)
+        {
+            char buf[Inet::kMaxIPAddressStringLength];
+            char * ipAddressOut = ipAddress[j].ToString(buf, sizeof(buf));
```
```suggestion
            char * ipAddressOut = ipAddress[j].ToString(buf);
```

bzbarsky-apple commented on src/lib/mdns/Resolver.h:
```diff
@@ -138,6 +141,66 @@ struct DiscoveredNodeData
+            char buf[Inet::kMaxIPAddressStringLength];
+            char * ipAddressOut = ipAddress[j].ToString(buf, sizeof(buf));
+            ChipLogDetail(Discovery, "IP Address #%d: %s", j + 1, ipAddressOut);
+            (void) ipAddressOut;
```
Does this allow the ToString call to be optimized out when detail logging is disabled?  If not, this should use the appropriate detail logging ifdefs instead.  Followup is fine for this, but please make sure it happens.

sharadb-amazon commented:
> > Does that have something different with chip-device-ctrl > discover
> > ## In TE4_TestPlanVerificationSteps_V5, G34
> > Put the device in commisioning mode, Verify device advertising its ability for commissioning by running 'discover' command in the python controller along with the QR code or '-v vendorID' .. etc look for discover command usage and pass appropriate param to have commissioner discover the commissionable node . Look for the device information in the below command.
> > chip-device-ctrl > discover -qr MT:YNJV7VSC00KA0648G00
> > Waiting for device responses...
> > [1626844356.667520][1256856:1256864] CHIP:DL: Avahi browse: cache exhausted
> > [1626844356.770602][1256856:1256864] CHIP:DL: Avahi browse: cache new
> > [1626844356.771830][1256856:1256864] CHIP:DL: Avahi browse: cache new
> > [1626844357.669397][1256856:1256864] CHIP:DL: Avahi browse: all for now
> > [1626844357.879577][1256856:1256864] CHIP:DL: Avahi resolve found
> > [1626844357.881913][1256856:1256864] CHIP:DL: Avahi resolve found
> > [1626844358.060376][1256856:1256856] CHIP:DIS: Commissionable Node 0
> > [1626844358.060452][1256856:1256856] CHIP:DIS: Host name: DCA6328D7072
> > [1626844358.060526][1256856:1256856] CHIP:DIS: Long discriminator: 3840
> > [1626844358.060591][1256856:1256856] CHIP:DIS: Vendor ID: 9050
> > [1626844358.060656][1256856:1256856] CHIP:DIS: Product ID: 65279
> > [1626844358.060748][1256856:1256856] CHIP:DIS: Additional Pairing 0
> > [1626844358.060808][1256856:1256856] CHIP:DIS: Commissioning Mode 1
> > [1626844358.060884][1256856:1256856] CHIP:DIS: Device Type 257
> > [1626844358.060950][1256856:1256856] CHIP:DIS: Device Name Test Bulb
> > [1626844358.061028][1256856:1256856] CHIP:DIS: Rotating Id 00007C5F6E176CD40F68685D100A1CF8A98B
> > [1626844358.061089][1256856:1256856] CHIP:DIS: Pairing Instruction
> > [1626844358.061160][1256856:1256856] CHIP:DIS: Pairing Hint 0x21
> > [1626844358.061219][1256856:1256856] CHIP:DIS: Address 0: 192.168.4.223
> > chip-device-ctrl >
> 
> I see, this feature is not supported by chip-tool, but do we really need a standalone tv-casting-app? Or cloud added a command in chip tool?

I think specific commands like sending a UDC request can be modeled as a chip-tool command as well. But this is an example of an end-to-end tv-casting-app which implements this whole commissioning flow: [InitiatingSetup.adoc#24-commissioner-discovery-from-an-on-network-device](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/rendezvous/InitiatingSetup.adoc#24-commissioner-discovery-from-an-on-network-device) and is then supposed to send TV specific casting commands (TBD). 

We will also port this over to separate Android/iOS apps later that can be used as reference by any content provider to implement commissioning of itself and casting in their respective apps.
